src/main/java/com/iota/iri/service/milestone/impl/LatestSolidMilestoneTrackerImpl.java

```diff
@@ -0,0 +1,264 @@
+
+    /**
+     * Holds the milestone index of the milestone that caused the repair logic to get started.<br />
+     */
+    private int errorCausingMilestoneIndex = Integer.MAX_VALUE;
+
+    /**
+     * Counter for the backoff repair strategy (see {@link #revertPrecedingMilestones(MilestoneViewModel)}.<br />
+     */
+    private int repairBackoffCounter = 0;
+
+    /**
+     * Creates a manager that keeps track of the latest solid milestones and that triggers the application of these
+     * milestones and their corresponding balance changes to the latest {@link Snapshot} by incorporating a background
+     * worker that periodically checks for new solid milestones.<br />
+     * <br />
+     * We simply store the passed in dependencies in their corresponding properties.<br />
+     *
+     * @param tangle Tangle object which acts as a database interface
+     * @param snapshotProvider manager for the snapshots that allows us to retrieve the relevant snapshots of this node
+     * @param milestoneService contains the important business logic when dealing with milestones
+     * @param latestMilestoneTracker the manager that keeps track of the latest milestone
+     * @param ledgerService the manager for
+     * @param messageQ ZeroMQ interface that allows us to emit messages for external recipients
+     */
+    public LatestSolidMilestoneTrackerImpl(Tangle tangle, SnapshotProvider snapshotProvider,
+            MilestoneService milestoneService, LedgerService ledgerService,
+            LatestMilestoneTracker latestMilestoneTracker, MessageQ messageQ) {
+
+        this.tangle = tangle;
+        this.snapshotProvider = snapshotProvider;
+        this.milestoneService = milestoneService;
+        this.ledgerService = ledgerService;
+        this.latestMilestoneTracker = latestMilestoneTracker;
+        this.messageQ = messageQ;
+    }
+
+    @Override
+    public void start() {
+        executorService.silentScheduleWithFixedDelay(this::latestSolidMilestoneTrackerThread, 0, RESCAN_INTERVAL,
+                TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void shutdown() {
+        executorService.shutdownNow();
+    }
+
+    /**
+     * {@inheritDoc}
+     * <br />
+     * In addition to applying the found milestones to the ledger state it also issues log messages and keeps the
+     * {@link LatestMilestoneTracker} in sync (if we happen to process a new latest milestone faster).<br />
+     */
+    @Override
+    public void checkForNewLatestSolidMilestones() throws MilestoneException {
+        try {
+            int currentSolidMilestoneIndex = snapshotProvider.getLatestSnapshot().getIndex();
+            if (currentSolidMilestoneIndex < latestMilestoneTracker.getLatestMilestoneIndex()) {
+                MilestoneViewModel nextMilestone;
+                while (!Thread.currentThread().isInterrupted() &&
+                        (nextMilestone = MilestoneViewModel.get(tangle, currentSolidMilestoneIndex + 1)) != null &&
```

Comment:
Since the latestMilestoneTracker currently scans all candidates at the start it can take a really long time to have the correct value - If we have find another solid milestone here I think it should be processed without waiting for the latestMilestoneTracker to advance to this milestone (thats also the reason why we have the method that syncs with the latestMilestoneTracker further down in the code).